### ISSUE TITLE
Add possibility to add custom namespaces

### ISSUE THREAD

**Issue by tbille** [canonical-web-and-design/konf#22]
In our team we have a custom namespace: automation. Currently konf limits the namespaces to we can deploy to (staging, production and demo)

Could we add the possibility to add custom namespaces to konf so we can use konf to deploy to our cool namespaces :)

**Comment by cristinadresch** [comment]
@jkfran @nottrobin thoughts on this?

**Comment by nottrobin** [comment]
Could we discuss our understanding of how we plan to use namespaces a bit?

I think this is an important design decision for Konf. Does it need a specific understanding of the `staging` to `production` pipeline? Or is it actually reasonable to have it just have some generic namespace support? If we want to support custom namespaces, are they a wholly separate thing from `staging` to `production`? How does `demo` fit into all this? Is that more of a pipeline thing or more similar to this `automation` thing?

And why do we need a separate namespace for `automation`? What's the problem being solved there?

Do you think this needs a meeting? Or should we just start and collaborate on a design doc?

**Comment by tbille** [comment]
@nottrobin the reason we want to create an `automation` namespace is to run our Cronjobs separately from the services to have a better separation of concerns.

Cronjobs don't run in our "classic" way: staging then production, we deploy them directly (eg. [snapcraft-poller-script](https://jenkins.canonical.com/webteam/job/snapcraft-poller-script-publish/)).

The workplace engineering tribe will start to run more and more automation to perform daily actions, generate reports, etc. in the form of cronjobs and we thought that to manage them it would be simpler to have them all in the same namespace.

We shared a document in our tribe review and in backend catchup. You can see all the details about our K8s configuration in this spec: https://docs.google.com/document/d/1rok-ImqYXp9c-YtGkiZYsluC4SHGor7fVYccNTonD_4/edit#

Let us know what are your thoughts.

**Comment by jkfran** [comment]
All our current cronjobs run on the production namespace at the moment. Why do you want it to be automation for yours? Also, we have IS monitoring the production namespace for us.

**Comment by sowasred2012** [comment]
@jkfran @nottrobin @tbille what's the latest on this? Can it be triaged?

**Comment by nottrobin** [comment]
At present this is just a discussion I think. @tbille does this have a specific deadline or anything? I don't think so right? In which case I don't think it should be triaged until we've reached a conclusion and what we actually want to do.

I would be good to know @tbille's thoughts about @jkfran's questions. Here are some of mine:

> Why do you want it to be automation for your cronjobs?

I actually think "cronjobs" might be a nice namespace, but curious to know @tbille's thinking.

> Cronjobs are already differentiated on k8s

This is a very valid point - what actual downsides and advantages are there to a separate namespace?

- do cronjobs ever appear together with running services? Is it actually tidier to separate them or are they separate anyway?
- do we want to share secrets? It's convenient, but are there any security reasons for separating them?

I'm not sure IS's monitoring matters much, I'm sure they could work with any structure. I feel they could do with thinking more carefully about how they should be monitoring cronjobs anyway.

**Comment by anthonydillon** [comment]
To be discussed in the infra guild meeting on the 28th of March.

**Comment by jpmartinspt** [comment]
@tbille do you want to explore Argo a little bit before we do this or decide on a namespace name that pleases everyone?

**Comment by nottrobin** [comment]
So here's my impression coming out of [the meeting](https://github.com/canonical-web-and-design/meeting-notes/issues/1027):

- cronjobs are a special case, don't follow the standard staging->production flow, and it probably makes sense to put them in their own namespace
- "automation" isn't a great name to describe cronjobs
- other types of "automation" are not yet well enough understood to say where they should go or what name best describes them

So we need to understand what we want to put in this namespace. We could either limit it to cronjobs and call it "cronjobs", or we should define clearly what else will go in it and find a name that fits well (Joao suggested "workflows" I think).

**Comment by mtruj013** [comment]
@jkfran @nottrobin @tbille can we add this as a topic for the infra guild?

**Comment by nottrobin** [comment]
@mtruj013 yes please. Can you add it? Notes are here https://docs.google.com/document/d/16Tm9nKedNJAyqwIVc1HgItpDCQaRRH9lFwfkWr-gT-M/edit#heading=h.ekwn5u8ydrla

**Comment by nottrobin** [comment]
We discussed this in the infrastructure meeting. Let's just make a namespace called "cronjobs"

**Comment by nottrobin** [comment]
^ what do we need to do to make this a reality? Do we need to do actual work in Konf? @jkfran @tbille

**Comment by jkfran** [comment]
> ^ what do we need to do to make this a reality? Do we need to do actual work in Konf? @jkfran @tbille

We need a bit of work on Konf and redeploy our current cronjobs on k8s. @tbille @goulinkh have been using automation on their k8s they will need to be prepared as well.
